src/ripple/app/tx/impl/Change.cpp

```diff
@@ -90,8 +90,42 @@ Change::preclaim(PreclaimContext const& ctx)
 
     switch (ctx.tx.getTxnType())
     {
-        case ttAMENDMENT:
         case ttFEE:
+            if (ctx.view.rules().enabled(featureXRPFees))
+            {
+                if (!ctx.tx.isFieldPresent(sfBaseFeeDrops) ||
+                    !ctx.tx.isFieldPresent(sfReserveBaseDrops) ||
+                    !ctx.tx.isFieldPresent(sfReserveIncrementDrops))
+                    return temMALFORMED;
+                // The transaction should only have one set of fields or the
+                // other.
+                if (ctx.tx.isFieldPresent(sfBaseFee) ||
+                    ctx.tx.isFieldPresent(sfReferenceFeeUnits) ||
+                    ctx.tx.isFieldPresent(sfReserveBase) ||
+                    ctx.tx.isFieldPresent(sfReserveIncrement))
+                    return temMALFORMED;
+            }
+            else
+            {
+                // The transaction format formerly enforced these fields as
+                // required. With featureXRPFees, those fields were made
+                // optional with the expectation that they won't be used once
```

Comment:
I may need to clarify the comment, but "optional" refers to the transaction format, changed here: https://github.com/XRPLF/rippled/pull/4247/files#diff-3e7bd959490f33226dff70ce2bfc3e957f2126a35476b2f5048a0e0328f2e77bL158-R162

At the object level, all of the fields are defined as optional, because that's the only real way I know of to have backward compatibility.

So, yes, they are forbidden in the branch above, if the amendment is enabled. In this block, they're actually required, so the code needs to check that they're all there.